tools/convert_bigfile_gadget_format3.py

```diff
@@ -0,0 +1,271 @@
+
+TODO: Support Black hole data arrays better.
+"""
+
+from __future__ import print_function
+import argparse
+import os
+import os.path
+import re
+import glob
+import bigfile
+import h5py
+import numpy as np
+
+
+class NameMaps(object):
+    """Maps names between HDF5 and Bigfile."""
+    def __init__(self):
+        #Map of (only changed) block names between HDF5 and bigfile snapshots.
+        self.hdf_to_bigfile_map = { "Coordinates" : "Position",
+                                    "Velocities": "Velocity", "Masses": "Mass",
+                                    "NeutralHydrogenAbundance": "NeutralHydrogenFraction",
+                                    "GFM_Metallicity": "Metallicity",
+                                    "ParticleIDs" : "ID",
+                                  }
+        #This requires python 2.7
+        self.bigfile_to_hdf_map = {v : k for (k,v) in self.hdf_to_bigfile_map.items()}
+        #Leave the metallicity array unchanged.
+        del self.bigfile_to_hdf_map["Metallicity"]
+
+    def get_bigfile_name(self, hdfname):
+        """Get the bigfile name from an HDF5 name."""
+        try:
+            return self.hdf_to_bigfile_map[hdfname]
+        except KeyError:
+            #Unrecognised names are assumed unchanged
+            return hdfname
+
+    def get_hdf5_name(self, bigfilename):
+        """Get the HDF5 name from a bigfile name."""
+        try:
+            return self.bigfile_to_hdf_map[bigfilename]
+        except KeyError:
+            #Unrecognised names are assumed unchanged
+            return bigfilename
+
+#Global name registry
+names = NameMaps()
+
+def write_hdf_header(bf, hdf5, nfiles, npart_file):
+    """Generate an HDF5 header from a bigfile header."""
+    head = hdf5.create_group("Header")
+    hattr = head.attrs
+    battr = bf["Header"].attrs
+    #As a relic from Gadget-1, the total particle numbers
+    #are written as two separate 32 bit integers.
+    hattr["NumPart_Total"] = np.uint32(battr["TotNumPart"] % 2**32)
+    hattr["NumPart_Total_HighWord"] = np.uint32(battr["TotNumPart"] // 2**32)
+    hattr["NumPart_ThisFile"] = np.int32(npart_file)
+    hattr["NumFilesPerSnapshot"] = np.int32(nfiles)
+    #Assume star formation implies the rest.
+    flag_sfr = np.int32(battr["TotNumPart"][4] > 0)
+    hattr["Flag_Sfr"] = flag_sfr
+    hattr["Flag_Cooling"] = flag_sfr
+    hattr["Flag_StellarAge"] = flag_sfr
+    hattr["Flag_Metals"] = flag_sfr
+    hattr["Flag_Feedback"] = 0
+    hattr["Flag_DoublePrecision"] = (bf["1/Position"].dtype == np.float64)
+    hattr["Flag_IC_Info"] = 0
+    hattr["Redshift"] = 1./battr["Time"] - 1
+    #Pass other keys through unchanged. We whitelist expected keys to avoid confusing Gadget.
+    hdfats = ["MassTable", "Time", "BoxSize", "Omega0", "OmegaLambda", "HubbleParam", "OmegaBaryon", "UnitLength_in_cm", "UnitMass_in_g", "UnitVelocity_in_cm_per_s"]
```

Comment:
It may be more verbose, but writing them explicitly may be easier to understand and maintain in such situation.

Reply:
if I write them explicitly I worry I will at some point copy-paste badly and write, eg:
battr["OmegaK"] = hattr["Omega0"]. But I can change it if you really like.

Reply:
Something to keep in mind in next round of refactoring.